// icbsc.js/dist/PrivateKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var icbsc_numeric_1 = require("./icbsc-numeric");
var PrivateKey = (function () {
    function PrivateKey(key) {
        this.key = key;
    }
    PrivateKey.fromElliptic = function (privKey, keyType) {
        if (keyType === void 0) { keyType = icbsc_numeric_1.KeyType.k1; }
        var privArray = privKey.getPrivate().toArray();
        return new PrivateKey({
            type: keyType,
            data: privKey.getPrivate().toBuffer(),
        });
    };
    PrivateKey.fromString = function (keyString) {
        var key = icbsc_numeric_1.stringToPrivateKey(keyString);
        return new PrivateKey(key);
    };
    PrivateKey.prototype.toElliptic = function (ecurve) {
        if (!ecurve) {
            if (this.key.type === icbsc_numeric_1.KeyType.r1) {
                ecurve = icbsc_numeric_1.r1Ec;
            }
            else if (this.key.type === icbsc_numeric_1.KeyType.k1) {
                ecurve = icbsc_numeric_1.defaultEc;
            }
            else if (this.key.type === icbsc_numeric_1.KeyType.sm2) {
                ecurve = icbsc_numeric_1.sm2Ec;
            }
        }
        return ecurve.keyFromPrivate(this.key.data);
    };
    PrivateKey.prototype.toString = function () {
        return icbsc_numeric_1.privateKeyToString(this.key);
    };
    PrivateKey.prototype.getType = function () {
        return this.key.type;
    };
    return PrivateKey;
}());
exports.PrivateKey = PrivateKey;
